### PR TITLE
Similarity api

### PR DESCRIPTION
**Pull request by MaxBittker**
This pr adds a new tab on the issue page that displays other similar issues. it's behind a feature flag that's off by default

#nochanges

### CODE REVIEW

File: src/sentry/templatetags/sentry_react.py [hunk 1]
```diff
@@ -95,10 +95,15 @@ def get_react_config(context):
...
         is_superuser = False
 
+    enabled_features = []
+
     if user:
         user = extract_lazy_object(user)
+    if not user.is_anonymous():
+        enabled_features = [project.organization.slug + ':' + project.slug + ':similarity-view'
+            for project in Project.objects.filter(organization=User.objects.get(id=user.id).get_orgs()) if
+            features.has('projects:similarity-view', project=project)]
```
Comment (MaxBittker) [inline src/sentry/templatetags/sentry_react.py]: @mattrobenolt is this query cool?

Comment (MaxBittker) [inline src/sentry/templatetags/sentry_react.py]: ted told me i might want get_for_user

Comment (mattrobenolt) [inline src/sentry/templatetags/sentry_react.py]: No, this is actually an O(n) of queries. I think there is a helper, but I'd have to look into it.

File: src/sentry/conf/server.py [hunk 2]
```diff
@@ -708,6 +708,7 @@ def create_partitioned_queues(name):
     'organizations:sso': True,
     'organizations:callsigns': True,
     'organizations:group-unmerge': False,
+    'projects:similarity-view': True,
```
Comment (billyvg) [inline src/sentry/conf/server.py]: Should this be False?

File: src/sentry/templatetags/sentry_react.py (hunk 1 above)
Comment (billyvg) [inline src/sentry/templatetags/sentry_react.py]: @mattrobenolt Can we revisit this?

File: src/sentry/static/sentry/app/components/pagination.jsx [hunk 3]
```diff
@@ -48,7 +48,7 @@ export default React.createClass({
     }
 
     return (
-      <div className="stream-pagination">
+      <div className="stream-pagination clearfix">
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/pagination.jsx]: Unsure if we depend on this behavior somewhere, but this seems more correct

File: src/sentry/templatetags/sentry_react.py (hunk 1 above)
Comment (mattrobenolt) [inline src/sentry/templatetags/sentry_react.py]: Ok, so the query itself isn't O(n). Not sure what influence I was under.

But this query is going to be scoped incorrectly. The problem is it's not being scoped by team membership. You're saying, "any organizations this user is a part of, get all of the projects." which is invalid without checking Team membership. You may be a part of an org, but not a part of all of the teams.

This is where `Project.objects.get_for_user` comes into play, but takes `team` as an argument.

So without Team being present here, this is a bad check.

**Review by mattrobenolt — CHANGES_REQUESTED** [review]

File: src/sentry/templatetags/sentry_react.py (hunk 1 above)
Comment (mattrobenolt) [inline src/sentry/templatetags/sentry_react.py]: And to be clear, there are relatively expensive ways we can do this. You can do like:

```
projects = set()
for organization in Organization.objects.get_for_user(user):
    for team in Team.objects.get_for_user(organization, user):
        projects.update(Project.objects.get_for_user(team, user, _skip_team_check=True))
```

But to me, this just feels like you're going against the grain since this isn't a pattern that we do anywhere else and we have nothing optimized for it.

File: src/sentry/static/sentry/app/stores/groupingStore.jsx [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+      // Update "checked" state for row
+      this.setStateForId(this.unmergeState, id, {
+        checked
+      });
+
+      this.triggerUnmergeState();
+    }
+  },
+
+  onUnmerge({groupId, loadingMessage, successMessage, errorMessage}) {
```
Comment (benvinegar) [inline src/sentry/static/sentry/app/stores/groupingStore.jsx]: Aside: one of the things I really dislike about Reflux is that you can't "search" for usage of `onUnmerge` – since this is a generated name derived from the `merge` action. At first I was really confused about where these methods were called until I remembered how Reflux works.

File: src/sentry/static/sentry/app/views/groupGrouping/groupGroupingView.jsx [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        endpoint: this.getEndpoint('hashes'),
+        dataKey: 'merged',
+        queryParams: this.props.location.query
+      },
+      {
+        endpoint: this.getEndpoint('similar'),
+        dataKey: 'similar',
+        queryParams: this.props.location.query
+      }
+    ]);
```
Comment (benvinegar) [inline src/sentry/static/sentry/app/views/groupGrouping/groupGroupingView.jsx]: Looks nice having API requests separated from view code.

File: src/sentry/static/sentry/app/views/groupGrouping/mergedItem.jsx [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+  onGroupingChange({unmergeState}) {
+    let {fingerprint} = this.props;
+    if (unmergeState) {
+      const stateForId = unmergeState.has(fingerprint) && unmergeState.get(fingerprint);
+      if (stateForId) {
+        Object.keys(stateForId).forEach(key => {
+          if (stateForId[key] !== this.state[key]) {
+            this.setState({
+              [key]: stateForId[key]
```
Comment (benvinegar) [inline src/sentry/static/sentry/app/views/groupGrouping/mergedItem.jsx]: Same comment re: avoiding deep nesting. I'd personally write this as:

```javascript
if (!unmergeState) return;

const stateForId = unmergeState.has(fingerprint) && unmergeState.get(fingerprint);
if (!stateForId) return;

Object.keys(stateForId).forEach(key => {
  if (stateForId[key] !== this.state[key]) return;

  this.setState({
    [key]: stateForId[key]
  });
});
```

... and I think most of the code follows a similar style. FWIW, [Martin Fowler feels similarly](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html).

File: src/sentry/templatetags/sentry_react.py [hunk 7]
```diff
@@ -92,10 +92,11 @@ def get_react_config(context):
         messages = []
         is_superuser = False
 
+    enabled_features = []
+
```
Comment (benvinegar) [inline src/sentry/templatetags/sentry_react.py]: Is moving this intentional? Doesn't seem to do anything, I'd just back this out.

File: src/sentry/static/sentry/app/components/splitLayout.jsx [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import React, {PropTypes} from 'react';
+import classNames from 'classnames';
+import SpreadLayout from './spreadLayout';
+
+const SplitLayout = ({children, className, responsive, ...props}) => {
```
Comment (benvinegar) [inline src/sentry/static/sentry/app/components/splitLayout.jsx]: What is the difference between `SplitLayout` and `SpreadLayout`? Maybe add a comment?

**Review by benvinegar — CHANGES_REQUESTED** [review]

File: src/sentry/static/sentry/app/stores/groupingStore.jsx (hunk 4 above)
Comment (billyvg) [inline src/sentry/static/sentry/app/stores/groupingStore.jsx]: Yeah I was also confused by this until reading the docs, it's hard to follow unless you know Reflux

File: src/sentry/static/sentry/app/stores/groupingStore.jsx [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+    let loadingIndicator = IndicatorStore.add(loadingMessage);
+
+    let promise = new Promise((resolve, reject) => {
+      api.request(`/issues/${groupId}/hashes/`, {
+        method: 'DELETE',
+        query: {
+          id: ids
+        },
+        success: (data, _, jqXHR) => {
+          IndicatorStore.remove(loadingIndicator);
```
Comment (billyvg) [inline src/sentry/static/sentry/app/stores/groupingStore.jsx]: Yeah I didn't feel too hot about this either. Do we have stores that create/dispatch actions?

File: src/sentry/static/sentry/app/stores/groupingStore.jsx [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+
+  setStateForId(map, id, newState) {
+    let state = (map.has(id) && map.get(id)) || {};
+    let mergedState = Object.assign({}, state, newState);
+    map.set(id, mergedState);
+    return mergedState;
+  },
+
+  // Fetches data
+  onFetch(toFetchArray) {
```
Comment (billyvg) [inline src/sentry/static/sentry/app/stores/groupingStore.jsx]: I'm not seeing the advantage of having this extra level of indirection. The action creators are coupled to one set of actions, wouldn't it be the same if they were implemented on the store?

That's not to say this shouldn't be refactored, but maybe as different action handlers `onFetch`, `onFetchStart`, `onFetchSuccess`, `onFetchError`.

Should the API calls to merge/unmerge also be action creators?

File: src/sentry/static/sentry/app/stores/groupingStore.jsx [hunk 11]
```diff
@@ -135,96 +174,73 @@ const GroupingStore = Reflux.createStore({
...
+        ({busy}) => busy
+      ) || [];
+
+      let canUnmerge =
+        this.unmergeList.size + 1 < this.mergedItems.length - lockedItems.length;
+      if (!canUnmerge) return;
+      this.unmergeList.add(id);
+      checked = true;
+
+      this.checkForRemainingUnmergeItem();
```
Comment (benvinegar) [inline src/sentry/static/sentry/app/stores/groupingStore.jsx]: This is much easier to follow now (for me) – thanks.

**Review by benvinegar — APPROVED** [review]
I'd still like to see a follow-up where we separate the `fetch`/`IndicatorStore` code from `groupingStore`. In the meantime let's move forward.

**Review by mattrobenolt — APPROVED** [review]

**Comment by MaxBittker** [comment]
🌮
